File: src/algorithms/arrays/difference-array/sources/difference-array.ts
// Difference Array — O(n + q) range updates via difference array and prefix sum reconstruction
function differenceArray(arrayLength: number, updates: number[][]): number[] {
  const diffArray: number[] = new Array(arrayLength + 1).fill(0); // @step:initialize
  const result: number[] = new Array(arrayLength).fill(0); // @step:initialize

  // Apply each range update [left, right, delta] to the difference array
  for (let updateIndex = 0; updateIndex < updates.length; updateIndex++) {
    const update = updates[updateIndex]!;
    const leftBound = update[0]!; // @step:visit
    const rightBound = update[1]!; // @step:visit
    const delta = update[2]!; // @step:visit
    diffArray[leftBound] += delta; // @step:compare
    if (rightBound + 1 < diffArray.length) {
      // @step:compare
      diffArray[rightBound + 1] -= delta; // @step:compare
    }
  }

  // Reconstruct result via prefix sum of the difference array
  let runningSum = 0; // @step:visit
  for (let scanIndex = 0; scanIndex < arrayLength; scanIndex++) {
    runningSum += diffArray[scanIndex]!; // @step:visit
    result[scanIndex] = runningSum; // @step:visit
  }

  return result; // @step:complete
}

File: src/algorithms/arrays/prefix-sum/difference-array/sources/difference-array.rs
// Difference Array — O(n + q) range updates via difference array and prefix sum reconstruction
fn difference_array(array_length: usize, updates: &[[i32; 3]]) -> Vec<i32> {
    let mut diff_array = vec![0i32; array_length + 1]; // @step:initialize
    let mut result = vec![0i32; array_length]; // @step:initialize

    // Apply each range update [left, right, delta] to the difference array
    for update_index in 0..updates.len() {
        let left_bound = updates[update_index][0] as usize; // @step:visit
        let right_bound = updates[update_index][1] as usize; // @step:visit
        let delta = updates[update_index][2]; // @step:visit
        diff_array[left_bound] += delta; // @step:compare
        if right_bound + 1 < diff_array.len() {
            // @step:compare
            diff_array[right_bound + 1] -= delta; // @step:compare
        }
    }

    // Reconstruct result via prefix sum of the difference array
    let mut running_sum = 0i32; // @step:visit
    for scan_index in 0..array_length {
        running_sum += diff_array[scan_index]; // @step:visit
        result[scan_index] = running_sum; // @step:visit
    }

    result // @step:complete
}
